29_tcp_chat/main.go

```diff
@@ -0,0 +1,166 @@
+package main
+
+import (
+	"bufio"
+	"net"
+	"strings"
+)
+
+const DEFAULT_ROOM = "global"
+
+var chat *Chat
+
+var help string = `
+HELP
+/comand params ...
+list of commands:
+	/setname username
+	/showrooms
+	/addroom name
+	/joinroom name
+`
+
+type Client struct {
+	id string
+	username string
+	room string
+	input chan string
+	output chan string
+	reader *bufio.Reader
+	writer *bufio.Writer
+}
+
+func (client *Client) Read() {
+	for {
+		line, _ := client.reader.ReadString('\n')
+
+		// check for a chat command
+		if strings.HasPrefix(line, "/") {
+			command := strings.Split(strings.TrimSpace(line), " ")
+			switch command[0] {
+			case "/setname":
+				client.username = command[1]
+			case "/showrooms":
+				for key := range chat.rooms {
+					client.Message(key + "\n")
+				}
+			case "/addroom":
+				if _, ok := chat.rooms[command[1]]; ok {
+					client.Message("Error: room already exists\n")
+				} else {
+					chat.rooms[command[1]] = CreateRoom()
+				}
+			case "/joinroom":
+				if _, ok := chat.rooms[command[1]]; ok {
+					delete(chat.rooms[client.room].clients, client.id)
+					client.room = command[1]
+					chat.rooms[client.room].clients[client.id] = client
+				} else {
+					client.Message("Error: room not exists\n")
+				}
+			default:
+				client.Message(help)
+			}
+		// add client info and forward further
+		} else {
+			chat.rooms[client.room].input <- client.id + "\n" + client.username + ": " + line
+		}
+	}
+}
+
+func (client *Client) Write() {
+	for data := range client.output {
+		client.Message(data)
+	}
+}
+
+func (client *Client) Message(msg string) {
+	client.writer.WriteString(msg)
+	client.writer.Flush()
+}
+
+func CreateClient(conn net.Conn) *Client {
+	client := &Client{
+		id: conn.RemoteAddr().String(),
```

Comment:
here we see that only `RemoteAddr().String()` is used, no need to copy entire `conn` object